### PR TITLE
Experiments with skill matching #51

### PR DESCRIPTION
**Pull request by masciotta02**
Test Matcher Skill

- example labels (VERB + customer satisfaction)
- cloud occupation (search skill into ESCO and create a matcher)

### CODE REVIEW

File: tests/test_skill_mathcer.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import spacy
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:1]: Fix file name;)

File: tests/test_skill_mathcer.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+from spacy.matcher import Matcher
+
+# Inizializza il modello di lingua di spacy
+nlp = spacy.load("en_core_web_sm")
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:5]: We use another spacy model in model.py

File: tests/test_skill_mathcer.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+# Crea un matcher
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:7]: Use English

File: tests/test_skill_mathcer.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                                    *** è rilevante a tal punto di utilizzare un altro pattern
+'ensure customer satisfaction',
+'guarantee customer satisfaction',
+'guaranteeing customer satisfaction',
+'promise customer satisfaction',
+'provide customer satisfaction',
+'to guarantee customer satisfaction']
+"""
+
+pattern = [
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:25]: You can start using only the simplest pattern.

Consider that in the end, patterns must be automatically generated.

File: tests/test_skill_mathcer.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+matcher.add("MatcherSkill", [pattern2])  # add pattern
+doc = nlp("I have a long experience in guaranteeing customer satisfaction ")
+doc1 = nlp("I have a long experience in providing satisfaction ")
+doc2 = nlp(
+    "I have a long experience in assure customer satisfaction and promise the customer satisfaction "
+)
+matches = matcher(doc)
+matches1 = matcher(doc1)
+matches2 = matcher(doc2)
+print("*************test_1*************")
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:61]: Create 3 different test functions, one for each matcher

File: tests/test_skill_mathcer.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+print("*************test_3*************")
+for match_id, start, end in matches2:
+    matched_span = doc2[start:end]
+    print(matched_span)
+
+# cloud position
+
+matched_sents = []
+
+pattern3 = [
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:78]: Where can I find test results?

File: tests/test_skill_mathcer.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    {"POS": "ADJ", "OP": "?"},
+    {"POS": "NOUN", "OP": "?"},
+    {"POS": "DET", "OP": "?"},
+    {"LOWER": "cloud"},
+    {"POS": "NOUN", "OP": "?"},
+]
+matcher.add("CloudSkill", [pattern3])
+doc = nlp(
+    "manage and maintain company IT infrastructure implement hybrid cloud solutions and design cloud architecture and plan migration cloud optimize"
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:102]: Talk with @Edocampa and try to create a test file together, based on the work that he's doing on the ner yaml file

**Review by ioggstream — CHANGES_REQUESTED** [review]
Good starting point!

Try to structure a test file like the others that we have in the repo.

Work together with your teammates,

File: tests/test_skill_mathcer.py (hunk 1 above)
Comment (masciotta02) [reply tests/test_skill_mathcer.py:1]: i don't understand this suggustesion

File: tests/test_skill_mathcer.py (hunk 2 above)
Comment (masciotta02) [reply tests/test_skill_mathcer.py:5]: i know, but i don't say how modify the code for implement the model that we use

File: tests/test_skill_mathcer.py (hunk 6 above)
Comment (masciotta02) [reply tests/test_skill_mathcer.py:78]: run the code .py and print the result

File: tests/test_skill_mathcer.py (hunk 1 above)
Comment (ioggstream) [reply tests/test_skill_mathcer.py:1]: s/mathcer/matcher/ # significa "sostituisci mathcer con matcher" cc: @Edocampa @Manu09011900

File: tests/test_skill_mathcer.py [hunk 8]
```diff
@@ -1,102 +1,100 @@
+import pytest
 import spacy
 from spacy.matcher import Matcher
 
-nlp = spacy.load("en_core_web_trf")
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: I replaced this global variable with a pytest.fixture that is reinitialized before every call.

File: tests/test_skill_mathcer.py [hunk 9]
```diff
@@ -1,102 +1,100 @@
...
-"""
-'assure customer satisfaction',
-'customer satisfaction guarantee',
-'ensure customer satisfaction',
-'guarantee customer satisfaction',
-'guaranteeing customer satisfaction',
-'promise customer satisfaction',
-'provide customer satisfaction',
-'to guarantee customer satisfaction']
-"""
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: don't need this anymore

File: tests/test_skill_mathcer.py [hunk 10]
```diff
@@ -1,102 +1,100 @@
...
-        "OP": "?",
-    },
-    {"POS": "DET", "OP": "?"},
-    {"TEXT": "customer", "OP": "?"},
-    {"TEXT": "satisfaction"},
-    {
-        "LEMMA": {"IN": ["guarantee", "assure", "ensure", "provide", "promise"]},
-        "OP": "?",
-    },
-]
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: Replaced with a dict()

File: tests/test_skill_mathcer.py [hunk 11]
```diff
@@ -1,102 +1,100 @@
...
-matcher.add("MatcherSkill", [pattern2])  # add pattern
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: now it's in the test code

File: tests/test_skill_mathcer.py [hunk 12]
```diff
@@ -1,102 +1,100 @@
...
-doc = nlp("I have a long experience in guaranteeing customer satisfaction ")
-doc1 = nlp("I have a long experience in providing satisfaction ")
-doc2 = nlp(
-    "I have a long experience in assure customer satisfaction and promise the customer satisfaction "
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: replaced with an array

File: tests/test_skill_mathcer.py [hunk 13]
```diff
@@ -1,102 +1,100 @@
...
-    matched_span = doc[start:end]
-    print(matched_span)
-print("*************test_2*************")
-for match_id, start, end in matches1:
-    matched_span = doc1[start:end]
-    print(matched_span)
-print("*************test_3*************")
-for match_id, start, end in matches2:
-    matched_span = doc2[start:end]
-    print(matched_span)
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: Replaced with test code

File: tests/test_skill_mathcer.py [hunk 14]
```diff
@@ -1,102 +1,100 @@
...
-            "IN": [
-                "design",
-                "deploy",
-                "plan",
-                "develop",
-                "manage",
-                "automate",
-                "implement",
-                "migrate",
-            ]
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py]: same as above

File: tests/test_skill_mathcer.py [hunk 15]
```diff
@@ -1,102 +1,100 @@
...
+@pytest.mark.parametrize("pattern", CUSTOMER_SATISFACTION_PATTERNS.values())
+def test_can_match_customer_satisfaction(nlp, text, pattern):
+    doc = nlp(text)
+    matcher = Matcher(nlp.vocab)
+    matcher.add("MatcherSkill", [pattern])
+    matches = matcher(doc)
+    for match_id, start, end in matches:
+        matched_span = doc[start:end]
+        print(matched_span)
+        assert matched_span.text == " ".join(pattern)
```
Comment (ioggstream) [inline tests/test_skill_mathcer.py:58]: for each pattern,
and for each text,
search for matches

File: tests/test_skill_pos_matcher.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    pattern = [{"LOWER": label.lower()}] if len(label) > 3 else [{"TEXT": label}]
+    patterns = [pattern]
+    altLabel = [sk["altLabel"]] if isinstance(sk["altLabel"], str) else sk["altLabel"]
+    for alt in altLabel:
+        candidate = generate_pattern_from_label(alt)
+        # Avoid duplicates.
+        if candidate not in patterns:
+            patterns.append(candidate)
+
+    return to_curie(id_), patterns
```
Comment (ioggstream) [inline tests/test_skill_pos_matcher.py:32]: This is the function generating the patterns for each skill

File: tests/test_skill_pos_matcher.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+        if nlp is None:
+            return [{"LOWER": label.lower()}]
+
+        nlp(label)
+        logging.warning(f"Generating pattern for: {label}")
+        # If the label is longer than 3 characters and has more than 3 words,
+        #   use spacy to generate the pattern.
+        raise NotImplementedError
+
+    yield f  #
```
Comment (ioggstream) [inline tests/test_skill_pos_matcher.py:62]: This is a feature that generates two generators. You need to write your code replacing the `raise NotImplementedError`.

The `nlp` parameter creates a different test for each nlp fixture

File: tests/test_skill_pos_matcher.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+
+@pytest.fixture
+def skills():
```
Comment (ioggstream) [inline tests/test_skill_pos_matcher.py:66]: returns only the skills related to cloud. This is to speed up the test.

File: tests/test_skill_pos_matcher.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    yield db.skills[db.skills.label.str.lower().isin(skills)]
+
+
+@pytest.fixture(params=(None, "merge_noun_chunks"))
+def nlp(request):
+    n = spacy.load("en_core_web_trf")
+    if request.param is not None:
+        n.add_pipe(request.param)
+    logging.warning(f"Using {n}")
+    yield n
```
Comment (ioggstream) [inline tests/test_skill_pos_matcher.py:91]: Creates 2 models, one vanilla, one with merge_noun_chunks to merge compound words such as "customer satisfaction" or "cloud services". We need to check what happens to lemmas in this case.

File: tests/test_skill_pos_matcher.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def test_esco_matcher(
+    skills,
+    generate_pattern_from_label,
+):
+    esco_matcher(skills, generate_pattern_from_label)
+    raise NotImplementedError
```
Comment (ioggstream) [inline tests/test_skill_pos_matcher.py:99]: This is the actual test: you need at first to safe the esco_matcher output in a json file :)

**Comment by masciotta02** [comment]
@ioggstream i try to add the code for yaml file test, but not work well... you can give me one suggestion. Thanks

**Comment by ioggstream** [comment]
Superseeded by #60

**Pull request closed without merging**